### PR TITLE
⚒Add typed contract hooks

### PR DESCRIPTION
**Pull request by Szymx95**
Screenshot 2021-12-17 at 11 52 54
Screenshot 2021-12-17 at 11 54 25
TypedContractCall will type check method and args

### CODE REVIEW

File: packages/core/src/model/types.ts [hunk 1]
```diff
@@ -1 +1,15 @@
+/* eslint-disable */
+import { Contract, ContractTransaction } from "ethers"
+
 export type Falsy = false | 0 | '' | null | undefined
+
+export type TypedContract = Contract & {
+    functions: Record<string, (...args: any[]) => any>
+}
+
+export type ContractFunctionNames<T extends TypedContract> = keyof { [P in keyof T['functions'] as ReturnType<T['functions'][P]> extends Promise<ContractTransaction> ? P : never] : void } & string
```
Comment (Szymx95) [inline packages/core/src/model/types.ts:10]: eslint for some reason finds a syntax error even though there is none. Probably something with version mismatch. Will have to check in future PR

File: .gitignore [hunk 2]
```diff
@@ -2,3 +2,4 @@ node_modules
 dist
 .DS_Store
 .idea
+packages/example/contracts
```
Comment (tt-marek) [inline .gitignore:5]: Noise?

File: packages/core/.prettierignore [hunk 3]
```diff
@@ -0,0 +1 @@
+src/model/types.ts
```
Comment (marekkirejczyk) [inline packages/core/.prettierignore:1]: Missing newline?

File: .gitignore (hunk 2 above)
Comment (Szymx95) [reply .gitignore:5]: Those are files generated with typechain 

> "typechain:generate": "yarn typechain --target=ethers-v5 ./src/abi/**/*.json --out-dir=./contracts"

Propably better not to include them in repo

Comment (tt-marek) [inline .gitignore:5]: Proposed directory: `packages/example/gen/types`

File: packages/example/package.json [hunk 4]
```diff
@@ -3,8 +3,9 @@
   "version": "0.2.10",
   "private": true,
   "scripts": {
+    "typechain:generate": "yarn typechain --target=ethers-v5 ./src/abi/**/*.json --out-dir=./contracts",
```
Comment (tt-marek) [inline packages/example/package.json:6]: -> `gen/types`

File: docs/source/core.rst [hunk 5]
```diff
@@ -107,6 +107,49 @@ It is recommended to use `useContractCall`_ where applicable instead of this met
 
 - ``calls: ChainCall[]`` - list of calls, also see `ChainCall`_. Calls need to be in the same order across component renders.
 
+.. _useTypedContractCall-label:
+
+useTypedContractCall
```
Comment (tt-marek) [inline docs/source/core.rst:112]: Can we merge `useTypedContractCall` and `useContractCall` into one function that accepts to different Call types as arguments?

**Pull request closed without merging**